docs: Add docs about maintenance mode (#1890)
nspcc-dev/neofs-node: docs/maintenance.md

# Maintenance mode of storage nodes

## Overview

Storage nodes go into maintenance mode (MM) while internal procedures are required
to troubleshoot storage service issues. Data on MM nodes MAY be temporarily
unavailable. Otherwise, such nodes do not differ from fully functional ones.

> Nodes go into MM on their own.

Reviewer: What did you mean here?

Reviewer 2: could please clarify some details:
- what does "Nodes go into MM on their own" mean?
- "Data on MM nodes MAY be temporarily unavailable" - if Data MAY be unable - what exactly affects on data availability?
- If user issues  "--status maintenance" command, is there any timeout to allow node return back to usual state?

Author: > Nodes go into MM on their own.

I mean MM can only be enabled by the node and no one else. I'll try to rephrase.

> what exactly affects on data availability?

The exact procedures performed on node side. Updated this place, pls check it out.

> is there any timeout to allow node return back to usual state?

Do you mean is there a timeout waiting for the state change?

Reviewer 2: I mean the timeout when the node will automatically back to online state.

Author: Currently no, the node must return to online state on its own.

## Reflection in the network map

To globally notify the network about the start of maintenance procedures, node
MAY set the corresponding state in the network map. The ability to set MM state
is determined by the global network setting. If MM is disallowed at the moment,

> node will not be able to set this state. Setting change is not retroactive.

Reviewer 3: ```suggestion
the node will not be able to set this state. Setting change is not retroactive.
```

> node will not be able to set this state. Setting change is not retroactive.

Reviewer 3: I don't quite get the `Setting change`. I would like a more explicit sentence.

Author: Paraphrased, check pls.

> To check the possibility of MM-switching at the moment, exec:

Reviewer 4: Mention `neofs-adm morph set-config` here?

```shell
$ neofs-cli netmap netinfo
...
Maintenance mode allowed: true
...
```

To switch the node to MM, exec:
```shell
$ control set-status --status maintenance
```

> To stop the maintenance, use same command with any other supported state.

Reviewer 3: ```suggestion
To stop the maintenance, use the same command but with any other supported state.
```

Reviewer 2: At least for the latest release, I am not able to see "Maintenance mode allowed" line while the MM is active.

It would be great to see "Maintenance mode allowed: true" while the MM is active and hide that line while we are in usual state.

Author: This setting didn't exist in previous release. I tested `netinfo` with `neofs-cli@v0.32.0`:
```
$ neofs-adm morph set-config MaintenanceModeAllowed=true
$ neofs-cli netmap netinfo
...
NeoFS network configuration (other)
  MaintenanceModeAllowed: 01
  SystemDNS: XXX
$ neofs-adm morph set-config MaintenanceModeAllowed=true
$ neofs-cli netmap netinfo
...
NeoFS network configuration (other)
  SystemDNS: XXX
$
```
Do you expect output like this?

Reviewer 2: I have tested and it seems those commands work fine for me. Could you please tell what exact "set-config MaintenanceModeAllowed=true" command affects? Only for displaying "Maintenance mode allowed: true" line?

Author: @vkarak1 the command changes config value in the running NeoFS network. From other side, `netinfo` shows current settings in the network.

> Note that node starts maintenance instantly, but network map state is changed

Reviewer 3: ```suggestion
Note that the node starts maintenance instantly, while the network map state is changed
```

asynchronously: Inner Ring receives switch requests, asserts allowance according

> to current network settings, and updates state of the candidate node for the

Reviewer 3: ```suggestion
to the current network settings, and updates the state of the candidate node for the
```

next network map. The change will take effect no earlier than the next epoch
in which a new version of the network map is released.

## Object service

Nodes under maintenance MAY fail local operations of the NeoFS Object API.

> Node maintained in current repository always denies all local object operations

Reviewer 3: ```suggestion
The node maintained in the current repository always denies all local object operations
```

# Maintenance mode of storage nodes

## Overview

Storage nodes turn to maintenance mode (MM) while internal procedures are required
to troubleshoot storage service issues. Data on MM nodes MAY be temporarily
unavailable depending on the procedures carried out at the site. Otherwise, such
nodes do not differ from fully functional ones. Nodes independently carry out
the procedure for turning on and off the MM.

## Reflection in the network map

To globally notify the network about the start of maintenance procedures, the node
MAY set the corresponding state in the network map. The ability to set MM state
is determined by the global network setting. If MM is disallowed at the moment,
the node will not be able to set this state. Network setting's change is not
retroactive: if MM becomes disallowed, the nodes with MM state are not excluded
from the network map.

To check the possibility of MM-switching at the moment, exec:
```shell
$ neofs-cli netmap netinfo
...
Maintenance mode allowed: true
...
```

To toggle the setting in the running NeoFS Sidechain, exec:
```shell
$ neofs-adm morph set-config MaintenanceModeAllowed=true|false
```

To switch the node to MM, exec:
```shell

> $ control set-status --status maintenance

Reviewer 4: `neofs-cli`?
